src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs

```diff
@@ -0,0 +1,159 @@
+                invocationFeatures.Set<IExecutionRetryFeature>(invocation);
+
+                context = _application.CreateContext(invocationFeatures, cancellationTokenSource.Token);
+                invocationFeatures.Set<IFunctionBindingsFeature>(new GrpcFunctionBindingsFeature(context, request, _outputBindingsInfoProvider));
+
+                if (_inputConversionFeatureProvider.TryCreate(typeof(DefaultInputConversionFeature), out var conversion))
+                {
+                    invocationFeatures.Set<IInputConversionFeature>(conversion!);
+                }
+
+                await _application.InvokeFunctionAsync(context);
+
+                var functionBindings = context.GetBindings();
+
+                foreach (var binding in functionBindings.OutputBindingData)
+                {
+                    var parameterBinding = new ParameterBinding
+                    {
+                        Name = binding.Key
+                    };
+
+                    if (binding.Value is not null)
+                    {
+                        parameterBinding.Data = await binding.Value.ToRpcAsync(_serializer);
+                    }
+
+                    response.OutputData.Add(parameterBinding);
+                }
+
+                if (functionBindings.InvocationResult is not null)
+                {
+                    TypedData? returnVal = await functionBindings.InvocationResult.ToRpcAsync(_serializer);
+                    response.ReturnValue = returnVal;
+                }
+
+                response.Result.Status = StatusResult.Types.Status.Success;
+            }
+            catch (Exception ex)
+            {
+                response.Result.Exception = ex.ToRpcException();
+                response.Result.Status = StatusResult.Types.Status.Failure;
+
+                if (ex.InnerException is TaskCanceledException or OperationCanceledException)
+                {
+                    response.Result.Status = StatusResult.Types.Status.Cancelled;
+                }
+            }
+            finally
+            {
+                _inflightInvocations.TryRemove(request.InvocationId, out var cts);
+
+                if (context is IAsyncDisposable asyncContext)
+                {
+                    await asyncContext.DisposeAsync();
+                }
+
+                (context as IDisposable)?.Dispose();
+            }
+
+            return response;
+        }
+
+        public bool TryCancel(string invocationId)
+        {
+            _inflightInvocations.TryGetValue(invocationId, out var cancellationTokenSource);
+
+            if (cancellationTokenSource is CancellationTokenSource)
+            {
+                try
+                {
+                    cancellationTokenSource?.Cancel();
+                    return true;
+                }
+                catch (ObjectDisposedException)
+                {
+                    // Do nothing, normal behavior
+                }
+                catch (Exception ex)
```

Comment:
Analyzers like to complain about broad exception catches. I like to add a `when (!ex.IsFatal())` where `IsFatal` is defined as:

``` CSharp
    /// <summary>
    /// Extensions for exceptions.
    /// </summary>
    public static class ExceptionExtensions
    {
        /// <summary>
        /// Determines if an exception is fatal, and should not be caught.
        /// </summary>
        /// <param name="exception">The exception to catch.</param>
        /// <returns>True if fatal, false otherwise (or if null).</returns>
        public static bool IsFatal(this Exception exception)
        {
            while (exception is not null)
            {
                if (exception.IsFatalCore())
                {
                    return true;
                }

                if (exception is AggregateException aggregate)
                {
                    // AggregateExceptions have a collection of inner exceptions, which may themselves be other
                    // wrapping exceptions (including nested AggregateExceptions). Recursively walk this
                    // hierarchy. The (singular) InnerException is included in the collection.
                    foreach (Exception innerException in aggregate.InnerExceptions)
                    {
                        if (innerException.IsFatal())
                        {
                            return true;
                        }
                    }

                    break;
                }
                else
                {
                    exception = exception.InnerException;
                }
            }

            return false;
        }

        private static bool IsFatalCore(this Exception exception)
        {
            return exception is OutOfMemoryException ||
                   exception is StackOverflowException || // can't actually be caught
                   exception is NullReferenceException || // this is a underlying code issue, we want to surface these.
                   exception is SEHException ||
                   exception is AccessViolationException || // can be caught, but means corrupted state.
                   exception is ThreadAbortException; // can be caught, but will always be re-thrown.
        }
    }
```

Reply:
We do this [in the host](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Extensions/ExceptionExtensions.cs), but looks like we don't have something similar in the worker repo.

I think probably worth doing this in a separate PR #1016